packages/@react-spectrum/utils/src/BreakpointProvider.tsx

```diff
@@ -0,0 +1,86 @@
+import React, {ReactNode, useContext, useEffect, useState} from 'react';
+import {useIsSSR} from '@react-aria/ssr';
+
+interface Breakpoints {
+  S?: number,
+  M?: number,
+  L?: number,
+  [custom: string]: number
+}
+
+interface IBreakpointContext {
+  matchedBreakpoints: string[]
+}
+
+export const BreakpointContext = React.createContext<IBreakpointContext>(null);
+BreakpointContext.displayName = 'BreakpointContext';
+
+// This is temporary values.
+export const DEFAULT_BREAKPOINTS = {S: 380, M: 768, L: 1024};
+
+interface BreakpointProviderProps {
+  children?: ReactNode,
+  matchedBreakpoints: string[]
+}
+
+export function BreakpointProvider(props: BreakpointProviderProps) {
+  let {
+    children,
+    matchedBreakpoints
+  } = props;
+  return (
+    <BreakpointContext.Provider
+      value={{matchedBreakpoints}} >
+      {children}
+    </BreakpointContext.Provider>
+  );
+}
+
+export function useMatchedBreakpoints(breakpoints: Breakpoints, mobileFirst: boolean): string[] {
+  // sort breakpoints
+  let entries = Object.entries(breakpoints).sort(([, valueA], [, valueB]) => mobileFirst ? valueB - valueA : valueA - valueB);
+  let breakpointQueries = entries.map(([, value]) => mobileFirst ?  `(min-width: ${value}px)` : `(max-width: ${value}px)`);
+
+  let supportsMatchMedia = typeof window !== 'undefined' && typeof window.matchMedia === 'function';
+  let getBreakpointHandler = () => {
+    let matched = [];
+    for (let i in breakpointQueries) {
+      let query = breakpointQueries[i];
+      if (window.matchMedia(query).matches) {
+        matched.push(entries[i][0]);
+      }
+    }
+    matched.push('base');
+    return matched;
+  };
+
+  let [breakpoint, setBreakpoint] = useState(() =>
+    supportsMatchMedia
+      ? getBreakpointHandler()
+      : ['base']
+  );
+
+  useEffect(() => {
+    if (!supportsMatchMedia) {
+      return;
+    }
+
+    let onResize = () => {
```

Comment:
Maybe you should call `onResize` before the resizing event to guarantee that you are on the correct screen?

Reply:
I believe you're asking for what this line handles? https://github.com/adobe/react-spectrum/pull/1684/files#diff-42004c28584ea3d877a7de08ad8962f5de549775ec22172c9cef438f0a85c3afR59
It'll only run the first time and does the same thing that `onResize` does. While this function will run in response to any future resize events.
Or am I missing a case?

Reply:
Oh, you right! I didn't see this before.